### ISSUE TITLE
All our base64_convert.[ch] helper functions use int instead of size_t

### ISSUE THREAD

**Issue by magnumripper** [magnumripper/JohnTheRipper#2132]
We will need to rewrite all those helper functions to use size_t for sizes/lengths in order to be usable for things like rar2john with files larger than a gigabyte or so (depending on encoding - a larger than 1 GB binary will become larger than 2 GB of hex, overflowing signed integer).

I started fixing it but ended up scrapping it and post-poning because some functions currently return negative for error, which will not be possible with size_t. So it's not quite as simple as changing some "int" to "size_t".

This is probably not a very urgent issue and it's not needed at all until we merge the "arbitrary length input lines" code discussed elsewhere. And anyway it should be fairly uncommon to end up needing to inline that much data. OTOH it's an easy fix once we figure out how to change that error return (perhaps simply return `__SIZE_MAX__`).

**Comment by jfoug** [comment]
Which helpers were you talking about?

base64_convert.[ch]   ?
misc.[ch] ?
common.[ch] ?


Somewhere else.

**Comment by magnumripper** [comment]
Like I said, base64_convert.[ch]. Most other function will never see really large data but personally I will probably try to use size_t more from now on. Things change rapidly.

**Comment by jfoug** [comment]
5b59b08

**Comment by magnumripper** [comment]
That commit has got to be broken. Like I wrote in the OP:

*I started fixing it but ended up scrapping it and post-poning** ***because some functions currently return negative for error, which will not be possible with size_t.*** *So it's not quite as simple as changing some "int" to "size_t".*

**Comment by magnumripper** [comment]
I guess you should revert this until you fix it properly.

**Comment by magnumripper** [comment]
No, looks like I'm wrong. Those functions still return int and the internals look good.

**Comment by magnumripper** [comment]
No I was right (lol). Here's one
```c
/*
 * if full string is HEX, then return is positive. If there is something
 * other than hex characters, then the return is negative but is the length
 * of 'valid' hex characters that start the string.
 *
 * NOTE, if the length is odd, then the 'full' length will not be returned.
 * the length returned will be length-1 since it would not be proper to try
 * to hex convert the last odd byte.
 */
static MAYBE_INLINE size_t _hexlen(const char *q, unsigned char dic[0x100])
{
	const char *s = q;
	size_t len = strlen(q);
	if (len&1) --len;

	while (dic[ARCH_INDEX(*q)] != 0x7F)
		++q;
	if ((size_t)(q - s)&1) --q;
	return (len == (size_t)(q - s)) ? (int)(q - s) : 0 - (int)(q - s);
}
```

**Comment by magnumripper** [comment]
Also, I would think more functions need fixes, eg. `int base64_valid_length()`.

**Comment by jfoug** [comment]
I will just back out common.c\h and look closer at base64-convert

**Comment by jfoug** [comment]
Please look over this:   bf9b4ee    common.ch will simply wait.  Like you mention, most longer work should be in base64_convert anyway.

**Comment by magnumripper** [comment]
Are you kidding? You don't seem to have a very good day today. PLEASE stop pushing commits to the main branch until you have something proper and ready for commit.

The very first row in base64_valid_length is `int len=0;`.

Hint 1: That's certainly one that needs to be size_t.
Hint 2: **We can't return `-len` because we return size_t.**

Don't be so trigger happy. Finish it, test it, and only then commit it. Feel free to push experiments to a topic branch though - that's what they're for. We have an embarrassing ratio of revert commits, makes us look like idiots.

**Comment by magnumripper** [comment]
I reverted both now.

**Comment by jfoug** [comment]
PR: 2168

**Comment by magnumripper** [comment]
Now we're talking.

**Comment by jfoug** [comment]
The extra field is not 'overly' used.  Too bad this was not C++, then we could have just made a default parameter, and the other 20 or so files would not have needed touched.

But the biggest thing was removing that error return code (not dual using the return).

NOTE, the functions in common.ch use long (not size_t).  They 'document' returning negative for partial base64 string. HOWEVER, they are unused, so we might either nuke them.  They really are 'not' needed, since they are just wrappers to the base64 stuff.

**Comment by magnumripper** [comment]
They were added in dbfdee2e. If they are unused we should definitely get rid of them.

**Comment by jfoug** [comment]
I believe this issue is now completed? correct?
